### DISCUSSION TITLE
Should retry on 4401

### DISCUSSION THREAD

**Discussion by pederbl** [enisdenjo/graphql-ws#559] [Q&A] [answered]
**Expected Behaviour**
I expected the client to retry connecting after the connection was terminated  by the server with a 4401. 

**Actual Behaviour**
But instead the retry is aborted and an error thrown, because the error code was 4401, because of the following code: (https://github.com/enisdenjo/graphql-ws/blob/63881c3372a3564bf42040e3f572dd74e41b2e49/src/client.ts#L834)

**Further Information**
Why disable reconnections for 4401 errors? When a token expires it is appropriate for the server to close the connection with a 4401. And, in this case it would be reasonable for the client to refresh the access token and retry the connection. 

My workaround now is to send a 4499 instead of 4401, which is bad semantics.

**Comment by enisdenjo** [comment, marked as answer]
Hey there! The [close code 4401 indicates that the client tried subscribing before the handshake's finished](https://the-guild.dev/graphql/ws/docs/enums/common.CloseCode#unauthorized) (before receiving a ConnectionAck message), meaning there's something wrong with the implementation of the client and should therefore not be retried.

> Executing operations is allowed only after the server has acknowledged the connection through the `ConnectionAck` message, if the connection is not acknowledged, the socket will be closed immediately with the event `4401: Unauthorized`.

_[GraphQL over WebSocket spec](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md)_

You may be confusing 4401 (Unauthorized) with 4403 (Forbidden), where the latter means that the implementation is OK but the user is not allowed to do something.

**Reply by pederbl** [reply]
Makes sense. Thanks!
